pkg/registers/registers.go

```diff
@@ -57,42 +70,138 @@ func (regs Registers) Find(id RegisterID) Register {
+	v, err := regs.forSerializationOBSOLETE()
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(v)
 }
 
 // UnmarshalJSON parses registers from JSON that was previously obtained using MarshalJSON
+// TODO: remove this, obsolete format
+// OBSOLETE
 func (regs *Registers) UnmarshalJSON(b []byte) error {
-	var jsonRegisters []registerJSON
+	var jsonRegisters registersForSerializationOBSOLETE
 	if err := json.Unmarshal(b, &jsonRegisters); err != nil {
-		return err
+		return fmt.Errorf("unable to un-JSON-ize registers: %w", err)
 	}
+	resultRegisters, err := jsonRegisters.ParseRegisters()
+	if err != nil {
+		return fmt.Errorf("unable to convert registers: %w", err)
+	}
+	*regs = resultRegisters
+	return nil
+}
+
+type registersForSerialization map[RegisterID]interface{}
 
-	resultRegisters := make(Registers, 0, len(jsonRegisters))
-	for _, jsonReg := range jsonRegisters {
-		reg, err := Unmarshal(jsonReg.ID, jsonReg.Value)
+func (m registersForSerialization) Registers() (Registers, error) {
+	resultRegisters := make(Registers, 0, len(m))
+	for regID, serializedValue := range m {
+		v, err := valueUnpack(regID, serializedValue)
+		if err != nil {
+			return nil, fmt.Errorf("unable to unpack value '%s': %w", serializedValue, err)
+		}
+		reg, err := New(regID, v)
 		if err != nil {
-			return err
+			return nil, err
 		}
 		resultRegisters = append(resultRegisters, reg)
 	}
-	*regs = resultRegisters
+	resultRegisters.Sort()
+	return resultRegisters, nil
+}
+
+// Sort sorts registers in some stable manner
+func (regs Registers) Sort() {
+	sort.Slice(regs, func(i, j int) bool {
+		a0 := regs[i].Address()
+		a1 := regs[j].Address()
+		if a0 != a1 {
+			return a0 < a1
+		}
+		return regs[i].ID() < regs[j].ID()
+	})
+}
+
+func (regs Registers) forSerialization() (registersForSerialization, error) {
+	m := make(registersForSerialization, len(regs))
+	for _, reg := range regs {
+		m[reg.ID()] = valueToHex(reg.Value())
+	}
+	return m, nil
+}
+
+// MarshalYAML converts a collection of registers into an YAML
+func (regs Registers) MarshalYAML() (interface{}, error) {
+	v, err := regs.forSerialization()
+	if err != nil {
+		return nil, fmt.Errorf("unable to prepare a serializable data structure: %w", err)
+	}
+
+	n := &yaml.Node{}
+	if err := n.Encode(v); err != nil {
+		return nil, fmt.Errorf("unable to encode to YAML: %w", err)
+	}
+	for _, regYaml := range n.Content {
+		regYaml.Style = yaml.TaggedStyle
+		regYaml.Tag = ""
+	}
+	return n, nil
+}
+
+// UnmarshalYAML parses registers from YAML that was previously obtained using MarshalYAML
+func (regs *Registers) UnmarshalYAML(value *yaml.Node) error {
+	var regsYaml registersForSerialization
+	if err := value.Decode(&regsYaml); err != nil {
+		return fmt.Errorf("unable to un-YAML-ize registers: %w", err)
+	}
+	result, err := regsYaml.Registers()
+	if err != nil {
+		return fmt.Errorf("unable to parse registers data: %w", err)
+	}
+	*regs = result
```

Comment:
Now the file with registers looks this way:
```
ACM_POLICY_STATUS: 0x12000d8e1
ACM_STATUS: 0xc0008000
TXT.DIDVID: 0x1b00c8086
TXT.DPR: 0x0
TXT.ERRORCODE: 0x0
TXT.ESTS: 0x0
TXT.HEAP.BASE: 0x0
TXT.HEAP.SIZE: 0x0
TXT.MLE.JOIN: 0x0
TXT.PUBLIC.KEY: 0x879a8f9cbf9e3d1d12dc9ad76dde34e6aa403664c739db347b858f0be033ae3a
TXT.SINIT.BASE: 0x0
TXT.SINIT.SIZE: 0x0
TXT.SPAD: 0x9800000080000000
TXT.STS: 0x12
TXT.VER.EMIF: 0x9d003000
TXT.VER.FSBIF: 0xffffffff
```
instead of old format:
```
[{"id":"ACM_POLICY_STATUS","value":"5dgAIAEAAAA="},{"id":"ACM_STATUS","value":"AIAAwAAAAAA="},{"id":"TXT.DPR","value":"AAAAAA=="},{"id":"TXT.ERRORCODE","value":"AAAAAA=="},{"id":"TXT.PUBLIC.KEY","value":"h5qPnL+ePR0S3JrXbd405qpANmTHOds0e4WPC+Azrjo="},{"id":"TXT.STS","value":"EgAAAAAAAAA="},{"id":"TXT.ESTS","value":"AA=="},{"id":"TXT.SPAD","value":"AAAAgAAAAJg="},{"id":"TXT.VER.FSBIF","value":"/////w=="},{"id":"TXT.VER.EMIF","value":"ADAAnQ=="},{"id":"TXT.DIDVID","value":"hoAMsAEAAAA="},{"id":"TXT.SINIT.BASE","value":"AAAAAA=="},{"id":"TXT.SINIT.SIZE","value":"AAAAAA=="},{"id":"TXT.MLE.JOIN","value":"AAAAAA=="},{"id":"TXT.HEAP.BASE","value":"AAAAAA=="},{"id":"TXT.HEAP.SIZE","value":"AAAAAA=="}]
```

Reply:
Marhsalling/unmarshalling registers from specific format should not be a method of Registers. Please make it as a funciton.

Reply:
Previous MarshalJSON should also be implemented as a function :)

Reply:
My logic is:
1. This is an intrinsic logic of `Registers`, why it should be defined outside of type `Registers`?
2. It is an usual practice in Go to define a method in such cases. I guess in all project I've seen it was defined as methods, not as functions.
3. When you define this as the method, you satisfy the [`Marshaler`](https://pkg.go.dev/encoding/json#Marshaler) and [`Unmarshaler`](https://pkg.go.dev/encoding/json#Unmarshaler) interfaces, which enables you to just use for example `json.Marshal` either directly on `Registers` or even on structure which includes `Registers`.

Could please you elaborate why it shouldn't be a method?

Reply:
>> Could please you elaborate why it shouldn't be a method?

Because without a method it could be even implemented in a different package.
A method introduces extra coupling that should be avoided if possible.

I forgot that golang requires types to implement interfaces, so it is ok to keep it as a method in this case